glib-macros/tests/test.rs

```diff
@@ -98,16 +89,19 @@ fn derive_gboxed_nullable() {
 
     let b = MyNullableBoxed(String::from("abc"));
     let v = b.to_value();
-    assert_eq!(&b, v.get::<&MyNullableBoxed>().unwrap().unwrap());
+    assert_eq!(&b, v.get::<Option<&MyNullableBoxed>>().unwrap().unwrap());
+    assert_eq!(b, v.get::<Option<MyNullableBoxed>>().unwrap().unwrap());
 
     let b = Some(MyNullableBoxed(String::from("def")));
-    let v = b.to_value();
+    let v = b.as_ref().to_value();
```

Comment:
This part is a bit of a change, but IMHO it makes more sense like this and allows more code than before.

Previously you could do ToValue::to_value(&Option<Object>)`, now you can do `ToValue::to_value(&Option<&Object>)`. You can easily get from `&Option<Object>` to `Option<&Object>` via `Option::as_ref()` but the other way around is not possible without cloning.

Reply:
One place where this was previously annoying was that e.g.

```rust
obj.set_property("foo", &Some(obj))
```

would consume `obj`, while now you can simply do `&Some(&obj)`.

Just as before, the following of course also still works

```rust
obj.set_property("foo", &obj)
```

Reply:
Taking `&Option<&Object>` seems a bit weird, and inconsistent with other gtk-rs function signatures, where it would be `Option<&Object>`.

Reply:
I agree but that's simply because `ToValue::to_value()` takes its argument by reference. You'd want it to take a `&Object`, for example, and not an `Object`.

Note that the first `&` in `set_property("foo", &Some(&obj))` is to turn it into a trait object.

Reply:
> to turn it into a trait object

Hm, doesn't look like `ObjectExt::set_property` uses a trait object; the function signature seems to use static dispatch:

```rust
fn set_property<'a, N: Into<&'a str>, V: ToValue>(
    &self,
    property_name: N,
    value: &V,   
) -> Result<(), BoolError>;
```

I suppose for it to work with `Option<&Object>`, `ToValue` would have to take `self` rather than `&self`, and be implemented for `&Object` rather than `Object`. Not sure if that would run into any issues.

Reply:
It would run into the issue that `Object` wouldn't be a `ValueType` :) The traits all would stop to fit together.

Reply:
This could now be

```rust
fn set_property<'a, N: Into<&'a str>, V: ToValue>(
    &self,
    property_name: N,
    value: V,   
) -> Result<(), BoolError>;
```
I.e. not `&V` but `V`. `ToValue` is implemented for `T` and `&T` and `&Option<T>` and `Option<&T>` so there are lots of combinations that can fit into that.

Reply:
With some reorganizing I got the advantages of all the above *and* the old code. Pushing soon.